python/ray/tune/experimental/output.py

```diff
@@ -0,0 +1,585 @@
+
+    rest -= hours * (60 * 60)
+    minutes = rest // 60
+
+    seconds = rest - minutes * 60
+
+    if days > 0:
+        running_for_str = f"{days:.0f} days, "
+    else:
+        running_for_str = ""
+
+    running_for_str += f"{hours:02.0f}:{minutes:02.0f}:{seconds:05.2f}"
+
+    return f"{current_time_dt:%Y-%m-%d %H:%M:%S}", running_for_str
+
+
+def _get_trials_by_state(trials: List[Trial]) -> Dict[str, List[Trial]]:
+    trials_by_state = collections.defaultdict(list)
+    for t in trials:
+        trials_by_state[t.status].append(t)
+    return trials_by_state
+
+
+def _infer_user_metrics(trials: List[Trial], limit: int = 4) -> List[str]:
+    """Try to infer the metrics to print out."""
+    result = list()
+    for t in trials:
+        if not t.last_result:
+            continue
+        for metric, value in t.last_result.items():
+            if metric not in DEFAULT_COLUMNS:
+                if metric not in AUTO_RESULT_KEYS:
+                    if type(value) in VALID_SUMMARY_TYPES:
+                        result.append(metric)
+
+            if len(result) >= limit:
+                return result
+    return result
+
+
+def _current_best_trial(trials: List[Trial], metric: str, mode: str):
+    if not trials:
+        return None, None
+
+    if not metric or not mode:
+        return None, metric
+
+    metric_op = 1.0 if mode == "max" else -1.0
+    best_metric = float("-inf")
+    best_trial = None
+    for t in trials:
+        if not t.last_result:
+            continue
+        metric_value = unflattened_lookup(metric, t.last_result, default=None)
+        if pd.isnull(metric_value):
+            continue
+        if not best_trial or metric_value * metric_op > best_metric:
+            best_metric = metric_value * metric_op
+            best_trial = t
+    return best_trial, metric
+
+
+@dataclass
+class _PerStatusTrialTableData:
+    trial_infos: List[List[str]]
+    more_info: str
+
+
+@dataclass
+class _TrialTableData:
+    header: List[str]
+    data: List[_PerStatusTrialTableData]
+
+
+def _max_len(value: Any, max_len: int = 20, wrap: bool = False) -> Any:
+    """Abbreviate a string representation of an object to `max_len` characters.
+
+    For numbers, booleans and None, the original value will be returned for
+    correct rendering in the table formatting tool.
+
+    Args:
+        value: Object to be represented as a string.
+        max_len: Maximum return string length.
+    """
+    if value is None or isinstance(value, (int, float, numbers.Number, bool)):
+        return value
+
+    string = str(value)
+    if len(string) <= max_len:
+        return string
+
+    if wrap:
+        # Maximum two rows.
+        # Todo: Make this configurable in the refactor
+        if len(value) > max_len * 2:
+            value = "..." + string[(3 - (max_len * 2)) :]
+
+        wrapped = textwrap.wrap(value, width=max_len)
+        return "\n".join(wrapped)
+
+    result = "..." + string[(3 - max_len) :]
```

Comment:
interesting, we keep the last part of the string and replace the prefix with ...?
is there a specific reason we do this?

Reply:
this is duplicated from `progress_reporter.py` - intend to keep it the same.